File: solutions/ruby/wordy/1/wordy.rb
class WordProblem
  OPERATIONS = {
    'plus' => -> (a, b) {a + b},
    'minus' => -> (a, b) {a - b},
    'multiplied by' => -> (a, b) {a * b},
    'divided by' => -> (a, b) {a / b}
  }.freeze

  def initialize(sentence)
    @sentence = sentence
  end
  
  def answer
    raise ArgumentError.new unless @sentence =~ /^What is (-?\d+) (?>(plus|minus|multiplied by|divided by) (-?\d+)\s?)+\?$/

    tokens = @sentence
      .scan(/-?\d+ (?>plus|minus|multiplied by|divided by)/)
      .to_a
      .map(&:split)
      .map{|number, *operator|
        [number.to_i, operator.join(' ')]
      }
      .push(@sentence.scan(/(-?\d+)\?/).at(0).at(0).to_i)
      .flatten

    res = 0
    while tokens.length != 1
      n1,operator,n2 = tokens.shift(3)
      res = OPERATIONS[operator].call(n1, n2)
      tokens.insert(0, res)
    end
    res
  end
end

File: solutions/python/wordy/1/wordy.py
import operator


def answer(question: str):
    print(question)
    if question.startswith('Who'):
        raise ValueError("unknown operation")

    if not question.startswith('What is ') or not question.endswith('?'):
        raise ValueError("syntax error")

    question = question.removeprefix('What is ')
    question = question.removesuffix('?')
    question = question.replace('plus', '+')
    question = question.replace('minus', '-')
    question = question.replace('multiplied by', '*')
    question = question.replace('divided by', '/')

    if any(unknown_operator in question for unknown_operator in ['cubed']):
        raise ValueError("unknown operation")

    if any(not char.removeprefix('-').isdigit() and char not in '+-*/' for char in question.split(' ')):
        raise ValueError("syntax error")

    operator_map = {
        '+': operator.add,
        '-': operator.sub,
        '*': operator.mul,
        '/': operator.floordiv
    }
    operator_fn = operator.add
    value, *other = question.split(' ')

    try:
        if len(other) == 0:
            return int(value)
        response = int(value)

        if response is None:
            raise ValueError("syntax error")

        while True:
            operator_str, value, *other = other
            operator_fn = operator_map.get(operator_str)
            if operator_fn is None:
                raise ValueError("syntax error")
            value = int(value)
            response = operator_fn(response, value)
            if len(other) == 0:
                break
    except Exception as exc:
        raise ValueError("syntax error") from exc

    return response
